src/wp-includes/option.php

```diff
@@ -782,16 +782,8 @@ function update_option( $option, $value, $autoload = null ) {
 	 */
 	$value = apply_filters( 'pre_update_option', $value, $option, $old_value );
 
-	/*
-	 * If the new and old values are the same, no need to update.
-	 *
-	 * Unserialized values will be adequate in most cases. If the unserialized
-	 * data differs, the (maybe) serialized data is checked to avoid
-	 * unnecessary database calls for otherwise identical object instances.
-	 *
-	 * See https://core.trac.wordpress.org/ticket/38903
-	 */
-	if ( $value === $old_value || maybe_serialize( $value ) === maybe_serialize( $old_value ) ) {
+	// If the new and old values are the same, no need to update.
+	if ( false !== $old_value && _is_equal_database_value( $old_value, $value ) ) {
```

Comment:
Why the `false` check here? If it's to see that no old value is stored in the database, that is unreliable, since registered settings may return another default than `false`.

Reply:
Is this an edge case where we cannot find any scenario in the old/new tests?

Reply:
See these [Plugins](https://wpdirectory.net/search/01HA7N5JVZKMCD54QKPFYMB0A2) and [Themes](https://wpdirectory.net/search/01HA7N7E8NSAK7WAX91449DJZJ) in the repository that make use of the `default_option_{$option}` filter.

**Note:** The above links don't show all extender usage. Not all plugins/themes are in the repository - There will be Premium/Self-hosted plugins and themes, or bespoke site-specific plugins and themes that make use of this filter.

If one of the filtered options isn't in the database, the filter will set the default value to something. Checking `false !== $old_value` would therefore be an inaccurate way of testing that an option doesn't exist in the database. It's a safe assumption that a filtered default value won't be `false`, as that's the original default value that's being overridden by this filter.

Reply:
The check using `default_option_{$option}` is in fact already part of the `update_option()` function, right after this logic here. I agree we should go with the approach @costdev you proposed above in https://github.com/WordPress/wordpress-develop/pull/5139/files#r1323397189.

@mukeshpanchal27 Can you implement that change please? We should then also add a few tests for that, e.g.:
* Call `update_option( 'something', false )` on an option that is not in the DB and ensure the value is still not stored in the DB afterwards.
* Call `update_option( 'something', 'default-value' )` after using the `default_option_something` filter to set the default to `'default-value'`, and ensure the value is still not stored in the DB afterwards.

Those tests should already pass in the current version of WordPress, i.e. without this PR. Would be great to do a quick test in another branch, where you _just_ add those tests but nothing else. If they pass without the new production code _and_ with the new production code, then we should be good.

Reply:
@mukeshpanchal27 These tests should do the trick to make sure the option isn't stored in the database.

```php
/**
 * Tests that update_option() does not store an option
 * that uses an unfiltered default value of (bool) false.
 *
 * @ticket 22192
 *
 * @covers ::update_option
 */
public function test_update_option_should_not_store_option_with_default_value_false() {
	global $wpdb;

	$option = 'update_option_default_false';
	update_option( $option, false );

	$actual = $wpdb->query(
		$wpdb->prepare(
			"SELECT option_name FROM $wpdb->options WHERE option_name = %s LIMIT 1",
			$option
		)
	);

	$this->assertSame( 0, $actual );
}

/**
 * Tests that update_option() does not store an option
 * that uses a filtered default value.
 *
 * @ticket 22192
 *
 * @covers ::update_option
 */
public function test_update_option_should_not_store_option_with_filtered_default_value() {
	global $wpdb;

	$option        = 'update_option_custom_default';
	$default_value = 'default-value';

	add_filter(
		"default_option_{$option}",
		static function() use ( $default_value ) {
			return $default_value;
		}
	);

	update_option( $option, $default_value );

	$actual = $wpdb->query(
		$wpdb->prepare(
			"SELECT option_name FROM $wpdb->options WHERE option_name = %s LIMIT 1",
			$option
		)
	);

	$this->assertSame( 0, $actual );
}
```

Reply:
Thanks @felixarntz and @costdev for the feedback. In https://github.com/WordPress/wordpress-develop/pull/5139/commits/bf1ef31f139f1d4c35abb0896d0d02cd5800c1da i have updated code and unit tests as suggested. Thanks!